### PR TITLE
Add game rule to manage diplomatic interaction between players

### PR DESCRIPTION
**Pull request by o01eg**
Uses set of allowed values to be extended in the future like in FreeCiv:

freeciv-d

### CODE REVIEW

File: Empire/EmpireManager.cpp [hunk 1]
```diff
@@ -226,6 +238,9 @@ void EmpireManager::HandleDiplomaticMessage(const DiplomaticMessage& message) {
     int sender_empire_id = message.SenderEmpireID();
     int recipient_empire_id = message.RecipientEmpireID();
 
+    if (GetGameRules().Get<std::string>("RULE_DIPLOMACY") == "RULE_DIPLOMACY_FORBIDDEN_FOR_ALL")
```
Comment (geoffthemedio) [inline Empire/EmpireManager.cpp]: String literals that are used as stringtable entry keys, if not being passed as a parameter to a function, should generally be wrapped in `UserStringNoOp` (or whatever it is, I can't check the spelling right now)

File: Empire/EmpireManager.cpp [hunk 2]
```diff
@@ -2,13 +2,25 @@
...
 #include <boost/filesystem/fstream.hpp>
 
 namespace {
+    void AddRules(GameRules& rules) {
+        // determine if diplomacy allowed
+        rules.Add<std::string>("RULE_DIPLOMACY", "RULE_DIPLOMACY_DESC",
+                               "MULTIPLAYER", "ALLOWED_FOR_ALL", false,
+                               DiscreteValidator<std::string>(std::set<std::string>({
+                                   "RULE_DIPLOMACY_ALLOWED_FOR_ALL",
+                                   "RULE_DIPLOMACY_FORBIDDEN_FOR_ALL"
```
Comment (geoffthemedio) [inline Empire/EmpireManager.cpp]: Might also make sense to wrap the literals here... not sure what standard practice for this would be as I don't think there are many cases of `DiscreteValidator` with strings...

File: Empire/EmpireManager.cpp [hunk 3]
```diff
@@ -2,13 +2,25 @@
...
 #include "../util/Directories.h"
+#include "../util/GameRules.h"
 #include "../util/Logger.h"
 #include "../util/XMLDoc.h"
 #include "../universe/Enums.h"
 
 #include <boost/filesystem/fstream.hpp>
 
 namespace {
+    void AddRules(GameRules& rules) {
```
Comment (geoffthemedio) [inline Empire/EmpireManager.cpp]: Why is this here and not in Diplomacy.cpp?

File: default/stringtables/en.txt [hunk 4]
```diff
@@ -1227,6 +1227,18 @@ Show detailed empires data
 RULE_SHOW_DETAILED_EMPIRES_DATA_DESC
 Show research and production, researched buildings, parts, and hulls of all empires.
 
+RULE_DIPLOMACY
+Manage diplomatic restrictions
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: Maybe "Allowed diplomacy"

File: Empire/EmpireManager.cpp (hunk 3 above)
Comment (o01eg) [reply Empire/EmpireManager.cpp]: Because it is used only here.

File: Empire/EmpireManager.cpp (hunk 2 above)
Comment (o01eg) [reply Empire/EmpireManager.cpp]: In the `CombatEvents.cpp` nothing is wrapped.

File: Empire/EmpireManager.cpp (hunk 3 above)
Comment (o01eg) [reply Empire/EmpireManager.cpp]: Moved code to `Dimplomacy.cpp`

File: default/stringtables/en.txt (hunk 4 above)
Comment (o01eg) [reply default/stringtables/en.txt]: Fixed.

File: Empire/EmpireManager.cpp (hunk 1 above)
Comment (o01eg) [reply Empire/EmpireManager.cpp]: Wrapped.

File: Empire/EmpireManager.cpp (hunk 2 above)
Comment (geoffthemedio) [reply Empire/EmpireManager.cpp]: Those aren't using `DiscreteValidator` though.

Looking over code a bit, it looks like in (almost?) all cases, calls to `OptionsDB::Add` wrap their string-literal parameters for option name in `UserStringNop`, but in no case I saw does a call to `GameRules::Add` do so. It probably should in all cases, though it doesn't need to be done so for this pull request specifically.

**Comment by geoffthemedio** [comment]
Some UI issues...
Diplomacy_Rule_Setup_Layout

**Comment by o01eg** [comment]
@geoffthemedio Test rule for discrete validator looks alike.

**Comment by geoffthemedio** [comment]
Test rule doesnt have the same text overflow issues for me.

**Comment by o01eg** [comment]
It looks like combobox issue. You could try to add characters to `en.txt` to reproduce it.

**Comment by geoffthemedio** [comment]
The issue is that the box is not wide enough for your text to fit, so one or both need to change size. Can you do that, or do you need / want me to?

**Comment by o01eg** [comment]
@geoffthemedio I think it's a dropdownlist issue: https://github.com/freeorion/freeorion/issues/2545

**Comment by geoffthemedio** [comment]
DropLists, including `StringRuleWidget` be made to now allow row contents to render outside them, but it doesn't really help much with this issue...
diplo_drop
I think your text is just too long for this space... even if the droplist is made a bit wider, more than one word is difficult to fit and all the text starts with the same word. Much wider and the description won't fit well...

Edit: I made the droplists (and the whole setup window and rules panel) wider, so the abbreviated text suggested below can fit.

File: default/stringtables/en.txt [hunk 5]
```diff
@@ -1227,6 +1227,18 @@ Show detailed empires data
...
 Show research and production, researched buildings, parts, and hulls of all empires.
 
+RULE_DIPLOMACY
+Allowed diplomacy
+
+RULE_DIPLOMACY_DESC
+Manages diplomatic restrictions between players. By default all interactions are allowed.
+
+RULE_DIPLOMACY_ALLOWED_FOR_ALL
+Diplomacy allowed for all
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: `All Allowed`

File: default/stringtables/en.txt [hunk 6]
```diff
@@ -1227,6 +1227,18 @@ Show detailed empires data
...
+
+RULE_DIPLOMACY_FORBIDDEN_FOR_ALL
+Diplomacy forbidden for all
```
Comment (geoffthemedio) [inline default/stringtables/en.txt]: `Always War` or similar

Comment (o01eg) [reply default/stringtables/en.txt]: I suppose it better to name `All forbidden` because for the same team #2542 it supposed to prevent breaking an alliance.

File: default/stringtables/en.txt (hunk 5 above)
Comment (o01eg) [reply default/stringtables/en.txt]: Fixed.

File: default/stringtables/en.txt (hunk 6 above)
Comment (o01eg) [reply default/stringtables/en.txt]: I suppose it better to name `All forbidden` because for the same team #2542 it supposed to prevent breaking an alliance.

Comment (geoffthemedio) [reply default/stringtables/en.txt]: It's not really clear from the name that it means "no changes after start" and not "always war between everyone"
